### PR TITLE
Add round trip time to metricset request

### PR DESCRIPTION
**Pull request by ruflin**
Add support for integer in templates

### CODE REVIEW

File: metricbeat/etc/fields.yml [hunk 1]
```diff
@@ -8,7 +8,28 @@ defaults:
...
+
+general:
+  type: group
+  description: >
+    General fields for each event
+  fields:
+    - name: response_time
+      type: long
+      description: >
+        Response time of the request in nano seconds
```
Comment (andrewkroh) [inline metricbeat/etc/fields.yml]: `s/nano seconds/nanoseconds/`

File: metricbeat/etc/fields.yml [hunk 2]
```diff
@@ -8,7 +8,28 @@ defaults:
...
+
+    - name: "@timestamp"
+      type: date
+      required: true
+      format: YYYY-MM-DDTHH:MM:SS.milliZ
+      example: 2015-01-24T14:06:05.071Z
+      description: >
+        The timestamp when the log line was read. The precision is in
```
Comment (andrewkroh) [inline metricbeat/etc/fields.yml]: "log line"?

File: metricbeat/helper/metricset.go [hunk 3]
```diff
@@ -52,12 +52,19 @@ func (m *MetricSet) Fetch() error {
 		go func(h string) {
 			defer m.Module.wg.Done()
 
+			starttime := time.Now()
 			event, err := m.MetricSeter.Fetch(m, h)
+			endtime := time.Now()
```
Comment (andrewkroh) [inline metricbeat/helper/metricset.go]: `elapsed := time.Since(starttime)` is the shorthand version of what you are doing with `time.Sub()`.

I wish Go had a precision clock for timing purposes (like Java's `System.nanoTime()`). By using wall-clock time your measurements are affected by clock changes (i.e. if NTP adjusts the wall-clock then your measurement is affected).

File: metricbeat/etc/fields.yml (hunk 1 above)
Comment (ruflin) [inline metricbeat/etc/fields.yml]: fixed

File: metricbeat/etc/fields.yml (hunk 2 above)
Comment (ruflin) [inline metricbeat/etc/fields.yml]: fixed

File: metricbeat/etc/fields.yml [hunk 4]
```diff
@@ -8,7 +8,28 @@ defaults:
...
 sections:
+  - ["general", "General event data"]
   - ["redis-info", "Redis Info Stats"]
+
+general:
+  type: group
+  description: >
+    General fields for each event
+  fields:
+    - name: response_time
```
Comment (andrewkroh) [inline metricbeat/etc/fields.yml]: In Packetbeat we have [`responsetime`](https://www.elastic.co/guide/en/beats/packetbeat/1.1/exported-fields-measurements.html#_responsetime) and it is measured in milliseconds. This will create an inconsistency.

File: metricbeat/helper/metricset.go (hunk 3 above)
Comment (ruflin) [inline metricbeat/helper/metricset.go]: Adjusted to the way you suggested.

Good point about the precision clock. Wasn't thinking of that but there doesn't see to be a better solution currently?

Comment (andrewkroh) [inline metricbeat/helper/metricset.go]: No, unfortunately there really isn't at the moment.

File: metricbeat/etc/fields.yml (hunk 4 above)
Comment (ruflin) [inline metricbeat/etc/fields.yml]: The problem is, that response times are sometimes lower then 1 millisecond, like 0.044 ms, means we would need a float. This could actually also be an issue in packetbeat I would assume?

Comment (ruflin) [inline metricbeat/etc/fields.yml]: About the naming: Yeah, not good to have to different ones, at the same time I think `response_time` would fit better into our naming scheme and `responsetime` is more because of legacy reasons?

Comment (andrewkroh) [inline metricbeat/etc/fields.yml]: I like `response_time` more than `responsetime`. And I like `response_time_nanos` even more.

Comment (ruflin) [inline metricbeat/etc/fields.yml]: The two concerns I have is that it sounds a little bit wiered (but I'm not the native one ;-)) and it somehow implies, there is also a response_time_millis

@tsg @monicasarbu Do you know if it was it ever an issue in packetbeat that responsetime is in milliseconds?

Comment (tsg) [inline metricbeat/etc/fields.yml]: The reason for millis is that nanos would have resulted into too human unfriendly values to show in Kibana. Best would be nanos + field formatter in Kibana, but it wasn't supported then and I'm not sure it is now. 

Since timelion is meant to be main consumer for this, we should check how good/bad that works in timelion at the moment. Long term, we plan to use mappings for defining the unit, right?

`response_time` is better than `responsetime` in a trivial way that IMO doesn't justify losing consistency or doing backwards incompatible changes.

Comment (ruflin) [inline metricbeat/etc/fields.yml]: I don't think using millis for response_time would make sense, as quite often the value would be 0. As far as I understood, field formatter the way we need it here is not supported yet, but could be added easily. 

@tsg For your last comment: So you would go with `responsetime` for consistency reason (we don't break any BC here)?

Other option: Going with `response_time` and change it to `response_time` in packetbeat in the next major version but having it in nano seconds too. A config flag would be used to enable `responsetime` as millis in packetbeat for the next major release and then removed. This would be my favorite option. I think also packetbeat could profit from using nanos.

Comment (tsg) [inline metricbeat/etc/fields.yml]: With BC I was referring to doing the rename in Packetbeat. I think @brandonmensing was suggesting that `request_time` or `latency` might be more correct?

I agree that we want nanos on the long term, but we should definitely check how nanos display in Timelion and Kibana, humans are just not that good at parsing nanoseconds. For example, this is how a filter for transactions taking more than 1 second would look like: `responsetime > 1000000000`. And the legend on the Y axis takes half the space. And the compression in Elasticsearch works a lot worse.  Putting that in balance with having precision below milliseconds (who really cares?).. It seem like at least a debatable choice to me.

Comment (ruflin) [inline metricbeat/etc/fields.yml]: I think there is a difference between packetbeat and metricbeat here. In packetbeat probably most of the time, the response_time is not below 1ms as the requests are happening between two servers. On metricbeat most of the requests should be to localhost and it happens very often, that a request takes less then 1ms. Having lots of 0 inside for response_time would be strange.

I can see `latency` and `request_time` in the context of packetbeat, not sure if it fits metricbeat as we actually "measure" the response time based on a timer.

Summary
* Packetbeat currently has `responsetime`. `latency` or `request_time` would probably be better. If  in ms or ns is not clear to me yet, depends also on Kibana.
* Metricbeat has `response_time`. Ns probably makes sense.

Sharing field names would be good but it seems like we will change Packetbeat in the future so I would suggest to go with `response_time` for metricbeat and "ignore" what we have right now in packetbeat.

Comment (brandonmensing) [inline metricbeat/etc/fields.yml]: The name sounds ok to me either way. I'm partial to response_time but don't let me opinion stop you :)

As for nanoseconds... I think we should not use them at all right now. It looks like Elasticsearch and Kibana both do not support nanoseconds (https://github.com/elastic/kibana/issues/2498 and https://github.com/elastic/elasticsearch/issues/10005). We should double check but if that's true, I think nanoseconds wouldn't be good right now.

As for the use cases, I have worked with people in past lives who needed nanosecond precision on service/server/app response times. That sounds nuts (and it is) but consider that they were working with the stock market :)

Comment (ruflin) [inline metricbeat/etc/fields.yml]: @brandonmensing Thanks for chiming in here. We are not taking nanoseconds timestamp but the diffs, so the numbers "should" be a reasonable small integer. If we come to the conclusion that response_time in ns is nothing useful, should we then provide response_time at all in ms if it is often 0? It is still nice to see things that take too long. I'm still +1 on the ns approach but I can see the point that it is probably overkill. Make it an optional flag to enable?

Comment (brandonmensing) [inline metricbeat/etc/fields.yml]: Ah, right. Since it's just a number, no problem with ES and Kibana's ms limitation. I'm ++ on a flag to enable ns. I think ms is a good default and I'm certain that some folks will need more precision.

Comment (ruflin) [inline metricbeat/etc/fields.yml]: For the flag I was referring to enabled requrest_time. If we mix ns and ms for the same type, we will probably end up with a mess.

Comment (brandonmensing) [inline metricbeat/etc/fields.yml]: Ok, I was assuming some other magic - that makes more sense!

Comment (ruflin) [inline metricbeat/etc/fields.yml]: @brandonmensing I made it now a config option

Comment (tsg) [inline metricbeat/etc/fields.yml]: Sorry, I thought about it a bit more and I'm not very happy with the outcome here. It's mostly my fault, too. 

Towards the beginning of the thread I thought we were talking about values returned by the systems. So like if Redis returns its average response time, we would put that in the `response_time` field, which would be a common field for all systems. Halfway through I realized it's about the queries that Metricbeat itself creates, and how long those take.

In this case, I think a more standard name is `round_trip_time` or `rtt`, because it doesn't only measure the response time, but the response time + the network delay between Metricbeat and the system. I'd prefer having a different name then what Packetbeat uses, because they really are different things. What do you think?

I think we can leave it always enabled and people can remove it with generic filtering (potentially conditionally if it's bigger than a threshold).

Comment (ruflin) [inline metricbeat/etc/fields.yml]: I really like the suggestion. I would go with `rtt` as it is shorter and I would say a common abbrevation? Ok going with nano seconds?

Comment (tsg) [inline metricbeat/etc/fields.yml]: Yes, good with nanos for this. <3

File: metricbeat/etc/metricbeat.template.json [hunk 5]
```diff
@@ -24,8 +24,56 @@
         "@timestamp": {
           "type": "date"
         },
+        "mysql-status": {
```
Comment (ruflin) [inline metricbeat/etc/metricbeat.template.json]: This is unrelated to this change and was updated because previously I forgot to run `make fields`.

**Comment by ruflin** [comment]
@tsg Updated with rtt

**Comment by tsg** [comment]
Thanks!
